### DISCUSSION TITLE
Lazy load data

### DISCUSSION THREAD

**Discussion by MisterOzzy** [reactivemarkets/react-financial-charts#472] [Q&A] [answered]
I have server which return OHLC data by time range. How can I subscribe or track when user move back chart by hand chart for get actual time range and send request to server and return specific data?

**Comment by markmcdowell** [comment, marked as answer]
There's 2 callbacks on the ChartCanvas to load data before or after the current window.

See:

https://github.com/reactivemarkets/react-financial-charts/blob/master/packages/core/src/ChartCanvas.tsx#L313

**Reply by MisterOzzy** [reply]
Could you explain more detailed? Because it's hard to understand how to retrieve time range which should be provided to return from callback and so on. or examples

**Reply by ivanmatiaspascual** [reply]
How can I do to this 2 callbacks to get it fired with zoom actions? not only pan actions

**Reply by n3xtgen** [reply]
Hi, I would really love someone to explain how to use the library for lazy loading previous candles for a user in a natural way. That is if user pans or zooms in graph the zoom and pan remains in the exact place and the graph just sends request for loading data and when the data is prepended to the candles array it adds them to the graph in the beginning.

What is currently happening to me is that when I add candles at the beginning of the array and the last candle in the graph is not visible at that moment. The graph shifts to the right because of adding those new candles at the beginning. If the last candle is visible, then the graph shifts to the right as much as there is space between the last candle and right edge of the graph.

Is there a way to make the graph work the same as TradingView in this regard? We would really appreciate at least some examples for this use case as there is otherwise almost no documentation. I have no idea how the graph should behave for various settings of xExtents for example.
